plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * Tests for change_title_placeholder_text().
+ *
+ * @package     spiralWebDb\CornerstoneTours\Tests\Unit
+ * @since       1.0.0
+ * @author      Robert Gadon <rgadon107>
+ * @link        https://github.com/rgadon107/cornerstone
+ * @license     GNU-2.0+
+ */
+
+namespace spiralWebDb\CornerstoneTours\Tests\Unit;
+
+use Mockery as m;
+use Brain\Monkey\Functions;
+use spiralWebDb\Cornerstone\Tests\Unit\Test_Case;
+use function spiralWebDb\CornerstoneTours\change_title_placeholder_text;
+
+/**
+ * Class Tests_ChangeTitlePlaceholderText
+ *
+ * @package spiralWebDb\CornerstoneTours\Tests\Unit
+ * @group   tours
+ * @group   admin
+ */
+class Tests_ChangeTitlePlaceholderText extends Test_Case {
+
+	/**
+	 * Instance of the post object for each test.
+	 *
+	 * @var Mockery
+	 */
+	protected $post;
+
+	/**
+	 * Prepares the test environment before each test.
+	 */
+	protected function setUp() {
+		parent::setUp();
+
+		require_once TOURS_ROOT_DIR . '/src/admin/edit-form-advanced.php';
+
+		$this->post          = m::mock( 'post' );
```

Comment:
Mockery mocks a class. `post` is not a class. Instead, it’s `m:::mock( WP_Post::class );`.

Reply:
@hellofromtonya I agree with you that the proper expression of the class to be mocked is `WP_Post::class`. However, when I pass that as an argument into `Mockery::mock()`, the unit test fails. 

The filter docblock (screenshot below) lists the second parameter of the filter as the post object. 

![wp-admin_edit-form-advanced](https://user-images.githubusercontent.com/3718603/78961072-b7d3c600-7ab5-11ea-9faf-424ca6da2212.png)

The PHP Manual says of the scope resolution operator ( double colon ): 

> The Scope Resolution Operator (also called Paamayim Nekudotayim) or in simpler terms, the double colon, is a token that allows access to static, constant, and overridden properties or methods of a class.

>When referencing these items from outside the class definition, use the name of the class.

>_As of PHP 5.3.0, it's possible to reference the class using a variable. The variable's value can not be a keyword (e.g. self, parent and static)._  (italics mine). 

The filter docblock uses the variable `$post` to represent `WP_Post::class` in the filter. When `post` is passed into Mockery as: 

>`$this->post  = m::mock( 'post' );`

the unit test passes. Unless you can come up with a workaround, I think we need to go with `$this->post  = m::mock( 'post' );`.

Reply:
Mockery mocks an instance of a class, i.e. objects. `'post'` is not class name or an instance of the object. Why?

1. The variable `$post` you noted above is within WordPress Core. Your test does not have access to that variable. 
2. The variable within core is bound to an instance of which class? `WP_Post`. 
3. Therefore, you want to mock the class and not the variable.

[Check out the documentation](http://docs.mockery.io/en/latest/reference/creating_test_doubles.html#classes-abstracts-interfaces) for mocking classes, interfaces, and abstracts.

Note: For this particular test, you could also pass in `null` instead of mocking `WP_Post`. Why? Think about it. 

- The callback function is not type hinting what `$post` parameter data type is. Yes, it's identified in the `@param` DocBlock, but not within the function's parameter list. 
- The test mocks `get_post_type()` function. Therefore, `$post` can be anything. Why? It's not used within the unit test.

What do I mean "type hinting" `$post`?

This would be type hinting a function parameter:

```php
function change_title_placeholder_text( $text, WP_Post $post ) {
```

How does this work?
When the callback is invoked, PHP checks the data type of what's being passed to the 2nd parameter. Huh? It checks to make sure the data type for `$post` is `WP_Post`. If no, a fatal error is thrown. 

It's best practice to type hint your code. Why? Increases the quality of your code. How? Your code is more reliable and robust (quality attributes) because it makes sure that it receives the type of data it expects.

Reply:
@hellofromtonya Thank you for responding back on this, and providing such a clear explanation.